pkg/supervisor/supervisor_unix.go

```diff
@@ -0,0 +1,126 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package supervisor
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+)
+
+const (
+	exitCheckInterval = 200 * time.Millisecond
+)
+
+// killPid signals SIGTERM to a PID and if it's still running after
+// s.TimeoutStop sends SIGKILL.
+func (s *Supervisor) killPid(pid int, check <-chan time.Time, deadline <-chan time.Time) error {
+	if s.KillFunction == nil {
+		s.KillFunction = syscall.Kill
+	}
+	// Kill the process pid
+	if deadline == nil {
+		deadlineTicker := time.NewTicker(s.TimeoutStop)
+		deadline = deadlineTicker.C
+		defer deadlineTicker.Stop()
+	}
+	if check == nil {
+		checkTicker := time.NewTicker(exitCheckInterval)
+		check = checkTicker.C
+		defer checkTicker.Stop()
+
+	}
+
+	// Using two tickers is not very elegant but makes testing easier...
+Loop:
+	for {
+		select {
+		case <-check:
+			err := s.KillFunction(pid, syscall.SIGTERM)
+			if err == syscall.ESRCH {
+				return nil
+			} else if err != nil {
+				return fmt.Errorf("failed to send SIGTERM to pid %d: %s", s.cmd.Process.Pid, err)
+			}
+		case <-deadline:
+			break Loop
+		}
+	}
+
+	err := s.KillFunction(pid, syscall.SIGKILL)
+	if err == syscall.ESRCH {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to send SIGKILL to pid %d: %s", s.cmd.Process.Pid, err)
+	}
+	return nil
+}
+
+// maybeKillPidFile checks kills the process in the pidFile if it's has
+// the same binary as the supervisor's and also checks that the env
+// `_KOS_MANAGED=yes`. This function does not delete the old pidFile as
+// this is done by the caller.
+// The tickers are used for testing purposes, otherwise set them to nil.
+func (s *Supervisor) maybeKillPidFile(check <-chan time.Time, deadline <-chan time.Time) error {
+	if s.ProcFSPath == "" {
+		s.ProcFSPath = "/proc"
+	}
+
+	pid, err := os.ReadFile(s.PidFile)
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to read pid file %s: %v", s.PidFile, err)
+	}
+
+	p, err := strconv.Atoi(strings.TrimSuffix(string(pid), "\n"))
+	if err != nil {
+		return fmt.Errorf("failed to parse pid file %s: %v", s.PidFile, err)
+	}
+
+	cmdline, err := os.ReadFile(filepath.Join(s.ProcFSPath, strconv.Itoa(p), "cmdline"))
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to read process %d cmdline: %v", p, err)
+	}
+
+	// only kill process if it has the expected cmd
+	cmd := strings.Split(string(cmdline), "\x00")
+	if cmd[0] != s.BinPath {
+		return nil
+	}
+
+	//only kill process if it has the _KOS_MANAGED env set
+	env, err := os.ReadFile(filepath.Join(s.ProcFSPath, strconv.Itoa(p), "environ"))
+	if err != nil {
+		return fmt.Errorf("failed to read process %d environ: %v", p, err)
+	}
```

Comment:
The process may be terminated after reading `cmdline`.
```suggestion
	if os.IsNotExist(err) {
		return nil
	} else if err != nil {
		return fmt.Errorf("failed to read process %d environ: %v", p, err)
	}
```